src/arkode/arkode_forcingstep.c

```diff
@@ -0,0 +1,414 @@
+                          at (tn, yn) = (tcur, ycur) from the end of the last
+                          step
+
+     ARK_FULLRHS_OTHER -> called elsewhere (e.g. for dense output)
+
+  In ForcingStep, we accumulate the RHS functions in ARK_FULLRHS_OTHER mode.
+  Generally, inner steppers will not have the correct yn when this function is
+  called and will not be able to reuse a function evaluation since their state
+  resets at the next ARKodeEvolve call.
+  ----------------------------------------------------------------------------*/
+static int forcingStep_FullRHS(const ARKodeMem ark_mem, const sunrealtype t,
+                               const N_Vector y, const N_Vector f,
+                               SUNDIALS_MAYBE_UNUSED const int mode)
+{
+  ARKodeForcingStepMem step_mem = NULL;
+  int retval = forcingStep_AccessStepMem(ark_mem, __func__, &step_mem);
+  if (retval != ARK_SUCCESS) { return (retval); }
+
+  const SUNStepper s0 = step_mem->stepper[0];
+  retval = s0->ops->fullrhs(s0, t, y, ark_mem->tempv1, ARK_FULLRHS_OTHER);
+  if (retval != 0)
+  {
+    arkProcessError(ark_mem, ARK_RHSFUNC_FAIL, __LINE__, __func__, __FILE__,
+                    MSG_ARK_RHSFUNC_FAILED, t);
+    return (ARK_RHSFUNC_FAIL);
+  }
+
+  const SUNStepper s1 = step_mem->stepper[1];
+  retval              = s1->ops->fullrhs(s1, t, y, f, ARK_FULLRHS_OTHER);
+  if (retval != 0)
+  {
+    arkProcessError(ark_mem, ARK_RHSFUNC_FAIL, __LINE__, __func__, __FILE__,
+                    MSG_ARK_RHSFUNC_FAILED, t);
+    return (ARK_RHSFUNC_FAIL);
+  }
+  N_VLinearSum(SUN_RCONST(1.0), f, SUN_RCONST(1.0), ark_mem->tempv1, f);
+
+  return ARK_SUCCESS;
+}
+
+/*---------------------------------------------------------------
+  This routine performs a single step of the forcing method.
+  ---------------------------------------------------------------*/
+static int forcingStep_TakeStep(const ARKodeMem ark_mem,
+                                sunrealtype* const dsmPtr, int* const nflagPtr)
+{
+  ARKodeForcingStepMem step_mem = NULL;
+  int retval = forcingStep_AccessStepMem(ark_mem, __func__, &step_mem);
+  if (retval != ARK_SUCCESS) { return (retval); }
+
+  *nflagPtr = ARK_SUCCESS; /* No algebraic solver */
+  *dsmPtr   = ZERO;        /* No error estimate */
+
+  const SUNStepper s0    = step_mem->stepper[0];
+  const sunrealtype tout = ark_mem->tn + ark_mem->h;
+  sunrealtype tret       = 0;
+  int stop_reason        = 0;
+
+  /* Evolve stepper 0 on its own */
+  SUNErrCode err = SUNStepper_Reset(s0, ark_mem->tn, ark_mem->yn);
+  if (err != SUN_SUCCESS) { return err; }
+  err = SUNStepper_SetStepDirection(s0, ark_mem->h);
+  if (err != SUN_SUCCESS) { return err; }
+  err = SUNStepper_SetStopTime(s0, tout);
+  if (err != SUN_SUCCESS) { return err; }
+  err = SUNStepper_Evolve(s0, ark_mem->tn, tout, ark_mem->ycur, &tret,
+                          &stop_reason);
+  if (err != SUN_SUCCESS) { return err; }
+  if (stop_reason < 0) { return stop_reason; }
```

Comment:
I'm not sure I understand why `stop_reason` was added and what values it should take. At least in the ARKStep implementation, it's just returning an error code:
https://github.com/LLNL/sundials/blob/d3ca3b8ad9034f71e2658303aaf990c2363e645c/src/arkode/arkode_arkstep.c#L3200-L3201

Reply:
I believe `stop_reason` was needed for the adjoint work as a factor in determining if a checkpoint should be written. We discussed having this be the return value instead of an output but that was problematic for some reason. I don't recall what the issue was. Maybe @balos1 remembers.

Reply:
> it's just returning an error code:

This is probably part of the reason. `ARKodeEvolve` returns an `int` value that's specific to ARKODE and not a `SUNErrCode`. Having it return a `SUNErrCode` would require changing return values for similar functions across the integrators to make them consistent which was undesirable.

Reply:
This has been addressed by mapping any non `SUN_SUCCESS` `SUNErrCode` to the ARKODE int error `ARK_SUNSTEPPER_ERR`. Then users can check `stepper->last_flag` to see the stepper-specific reason for the failure. I'll mark as resolved, but reopen or create a new comment if needed.